builder/image_solver.go

```diff
@@ -289,12 +269,38 @@ func (s *localRegistryImageSolver) SolveImage(ctx context.Context, mts *states.M
 		close(errChan)
 	}()
 
-	result := &states.ImageSolverResult{
-		ImageName:   dockerTag,
+	return &states.ImageSolverResults{
 		ResultChan:  resultChan,
 		ErrChan:     errChan,
 		ReleaseFunc: closer,
+	}, nil
+}
+
+func (m *multiImageSolver) addRefToResult(ctx context.Context, gwClient gwclient.Client, res *gwclient.Result, imageDef *states.ImageDef, idx int) error {
+	var saveImage = imageDef.MTS.Final.LastSaveImage()
+	imgJSON, err := json.Marshal(saveImage.Image)
+	if err != nil {
+		return errors.Wrap(err, "image json marshal")
 	}
 
-	return result, nil
+	if !strings.Contains(imageDef.ImageName, ":") {
+		imageDef.ImageName += ":latest"
+	}
+
+	ref, err := llbutil.StateToRef(ctx, gwClient, saveImage.State, true, imageDef.MTS.Final.PlatformResolver, m.cacheImports.AsMap())
+	if err != nil {
+		return errors.Wrap(err, "initial state to ref conversion")
+	}
+
+	refKey := fmt.Sprintf("image-%s", imageDef.ImageName)
+	refPrefix := fmt.Sprintf("ref/%s", refKey)
+	res.AddRef(refKey, ref)
+
+	localRegPullID := fmt.Sprintf("sess-%s/%s", gwClient.BuildOpts().SessionID, imageDef.ImageName)
+	res.AddMeta(fmt.Sprintf("%s/export-image-local-registry", refPrefix), []byte(localRegPullID))
+	res.AddMeta(fmt.Sprintf("%s/%s", refPrefix, exptypes.ExporterImageConfigKey), imgJSON)
+	res.AddMeta(fmt.Sprintf("%s/image.name", refPrefix), []byte(imageDef.ImageName))
+	res.AddMeta(fmt.Sprintf("%s/image-index", refPrefix), []byte(strconv.Itoa(idx)))
+
+	return nil
```

Comment:
I like the encapsulation of this logic in a nice little function.

Reply:
Thanks!